src/Application.php

```diff
@@ -45,8 +98,32 @@ public function middleware($middlewareQueue) : MiddlewareQueue
             ]))
 
             // Add routing middleware.
-            ->add(new RoutingMiddleware($this));
+            ->add(new RoutingMiddleware($this))
+
+            // Csrf Middleware
+            ->add($this->csrfMiddleware());
 
         return $middlewareQueue;
     }
+
+    /**
+     * Get internal Csrf Middleware
+     *
+     * @return CsrfProtectionMiddleware
+     */
+    protected function csrfMiddleware() : CsrfProtectionMiddleware
+    {
+        // Csrf Middleware
+        $csrf = new CsrfProtectionMiddleware();
+            // Token check will be skipped when callback returns `true`.
+        $csrf->whitelistCallback(function ($request) {
+            $actions = (array)Configure::read(sprintf('CrsfExceptions.%s', $request->getParam('controller')));
```

Comment:
Maybe we could avoid CSRF skip using `X-CSRF-Token` in ajax request, see https://book.cakephp.org/3.0/en/controllers/middleware.html#csrf-protection-and-ajax-requests

I didn't understand if it is our case :upside_down_face: